sidechain/consensus/aura/src/peer_block_sync.rs

```diff
@@ -0,0 +1,193 @@
+where
+	ParentchainBlock: ParentchainBlockTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+	BlockImporter: BlockImport<ParentchainBlock, SignedSidechainBlock>,
+	BlockProductionSuspender: SuspendBlockProduction + IsBlockProductionSuspended,
+{
+	pub fn new(
+		importer: Arc<BlockImporter>,
+		block_production_suspender: Arc<BlockProductionSuspender>,
+	) -> Self {
+		PeerBlockSync { importer, block_production_suspender, _phantom: Default::default() }
+	}
+
+	fn sync_blocks_from_peer(
+		&self,
+		_last_known_block_hash: BlockHash,
+		last_imported_parentchain_header: &ParentchainBlock::Header,
+	) -> Result<()> {
+		// TODO fetch blocks from peer
+
+		let blocks_to_import: Vec<SignedSidechainBlock> = Vec::new();
+		for block_to_import in blocks_to_import {
+			if let Err(e) =
+				self.importer.import_block(block_to_import, last_imported_parentchain_header)
+			{
+				// If we encounter an error here, we have to abort the sync process, no way to recover.
+				error!("Error while importing a block during peer sync process, cannot recover, have to abort");
+				return Err(e)
+			}
+		}
+
+		Ok(())
+	}
+}
+
+impl<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+	SyncBlockFromPeer<ParentchainBlock::Header, SignedSidechainBlock>
+	for PeerBlockSync<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+where
+	ParentchainBlock: ParentchainBlockTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+	BlockImporter: BlockImport<ParentchainBlock, SignedSidechainBlock>,
+	BlockProductionSuspender: SuspendBlockProduction + IsBlockProductionSuspended,
+{
+	fn attempt_block_sync(
+		&self,
+		sidechain_block: SignedSidechainBlock,
+		last_imported_parentchain_header: &ParentchainBlock::Header,
+	) -> Result<()> {
+		// In case block production is suspended, we don't import any blocks.
+		// In the future we might want to cache the blocks here, so they can be imported later.
+		if self.block_production_suspender.is_suspended().unwrap_or_default() {
+			warn!("Sidechain block won't be imported, since block production is suspended");
+			return Ok(())
+		}
+
+		// Attempt to import the block - in case we encounter an ancestry error, we go into
+		// peer fetching mode to fetch sidechain blocks from a peer and import those first.
+		if let Err(Error::BlockAncestryMismatch(_block_number, block_hash, _)) = self
+			.importer
+			.import_block(sidechain_block.clone(), last_imported_parentchain_header)
+		{
+			warn!("Got ancestry mismatch error upon block import. Attempting to sync missing blocks from peer");
+
+			// TODO need a 'finally' (or on-drop) here, to always resume block production ?
+
+			// Suspend block production while we sync blocks from peer.
+			self.block_production_suspender.suspend()?;
+
+			self.sync_blocks_from_peer(block_hash, last_imported_parentchain_header)?;
+
+			// Second attempt to import the original gossiped sidechain block
```

Comment:
Is that really necessary? We will most likely import this block when fetching the blocks from our peer. Or do you think we fetch faster than the block is included in our peers storage?

Reply:
Yes maybe - I can remove it again for now. I think we'll have to test it for a bit to see if it's necessary.